compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java

```diff
@@ -2769,6 +2805,146 @@ BType getTypeForUnionTypeMembersAssignableToType(BUnionType unionType, BType tar
         }
     }
 
+    BType getTypeIntersectedWithMapType(BMapType mapType, BType targetType, SymbolEnv env) {
+        if (targetType.tag != TypeTags.MAP && targetType.tag != TypeTags.RECORD) {
+            return symTable.semanticError;
+        }
+        if (targetType.tag == TypeTags.MAP) {
+            BType intersectionConstraintType = getTypeIntersection(mapType.constraint,
+                    ((BMapType) targetType).constraint, env);
+            if (intersectionConstraintType == symTable.semanticError) {
+                return symTable.semanticError;
+            }
+            return new BMapType(TypeTags.MAP, intersectionConstraintType, null);
+        }
+        return getTypeIntersectedWithRecordType((BRecordType) targetType, mapType, env);
+    }
+
+    BType getTypeIntersectedWithRecordType(BRecordType recordType, BType targetType, SymbolEnv env) {
+        if (targetType.tag != TypeTags.MAP && targetType.tag != TypeTags.RECORD) {
+            return symTable.semanticError;
+        }
+        if (targetType.tag == TypeTags.MAP) {
+            BRecordType intersectionRecordType = createAnonymousRecord(env);
+            LinkedHashMap<String, BField> fields = new LinkedHashMap<>();
+            for (String key : recordType.fields.keySet()) {
+                BType intersectionKeyType = getTypeIntersection(recordType.fields.get(key).type,
+                        ((BMapType) targetType).constraint, env);
+                if (intersectionKeyType == symTable.semanticError) {
+                    return symTable.semanticError;
+                }
+                BVarSymbol fieldSymbol = new BVarSymbol(0, names.fromString(key), env.enclPkg.symbol.pkgID,
+                        intersectionKeyType, intersectionRecordType.tsymbol, recordType.fields.get(key).pos,
+                        COMPILED_SOURCE);
+                BField field = new BField(names.fromString(key), recordType.fields.get(key).pos, fieldSymbol);
+                fields.put(key, field);
+            }
+            intersectionRecordType.fields = fields;
+
+            if (recordType.restFieldType != null) {
+                BType intersectionRestType = getTypeIntersection(recordType.restFieldType,
+                        ((BMapType) targetType).constraint, env);
+                if (intersectionRestType == symTable.semanticError) {
+                    intersectionRecordType.sealed = true;
+                }
+                intersectionRecordType.restFieldType = intersectionRestType;
```

Comment:
Is setting `intersectionRestType` correct even if `intersectionRestType` is `symTable.semanticError`? Shouln't it be `symTable.noType` if it is an exclusive record?

Reply:
Fixed